### PR TITLE
hmac encoded csrf token

### PR DESCRIPTION
**Pull request by willemvermeer**
Hi, this is a first iteration of a PR to resolve issue #77 
I have provided a minimal implementation and am open to suggestions for improvement. For example, I decided to reuse the existing server secret to hash the timestamp, instead of defining a second server secret for this new purpose (personally I think it can do no harm to reuse the same server secret, but feel free to disagree, in which case we'll have to add config changes for an additional server secret).
Best,
Willem

### CODE REVIEW

File: core/src/main/java/com/softwaremill/session/javadsl/CsrfDirectives.scala [hunk 1]
```diff
@@ -12,7 +12,7 @@ import com.softwaremill.session.CsrfCheckMode
 trait CsrfDirectives {
 
   def randomTokenCsrfProtection[T](checkMode: CsrfCheckMode[T], inner: Supplier[Route]): Route = RouteAdapter {
```
Comment (adamw) [inline core/src/main/java/com/softwaremill/session/javadsl/CsrfDirectives.scala]: we probably should have both `randomTokenCsrfProtection` (for backwards compatibility) and `hmacTokenCsrfProtection` method in the java api as well?

File: core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala [hunk 2]
```diff
@@ -36,6 +40,9 @@ trait CsrfDirectives {
     }
   }
 
+  // for backward compatibility
+  def randomTokenCsrfProtection[T](checkMode: CsrfCheckMode[T]): Directive0 = hmacTokenCsrfProtection(checkMode)
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala:44]: maybe we should add an `@deprecated` here

File: core/src/main/scala/com/softwaremill/session/SessionManager.scala [hunk 3]
```diff
@@ -84,7 +84,19 @@ trait CsrfManager[T] {
   def config: SessionConfig
 
   def tokenInvalidRejection = AuthorizationFailedRejection
-  def createToken(): String = SessionUtil.randomString(64)
+  def createToken(): String = {
+    val timestamp = System.currentTimeMillis().toString
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/session/SessionManager.scala]: the othre managers have an overridable `def nowMillis: Long` (for testing). Would be good to do the same thing here as well

File: core/src/main/scala/com/softwaremill/session/SessionManager.scala [hunk 4]
```diff
@@ -84,7 +84,19 @@ trait CsrfManager[T] {
...
+    val hmac = generateHmac(timestamp)
+    s"$hmac$timestamp"
+  }
+  def validateToken(token: String): Boolean = {
+    val len = System.currentTimeMillis().toString.length
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/session/SessionManager.scala]: I suppose we are assuming that the timestamp will always be `13` digits? Or rather, that between invocations of `create` and `validate` no extra digit will be added - unlikely, but possible ;). If we make this assumption, I think I would document it & add a constant, or use some other format which allows separating timestamp & hmac is a predictable way

File: core/src/main/scala/com/softwaremill/session/SessionManager.scala [hunk 5]
```diff
@@ -84,7 +84,19 @@ trait CsrfManager[T] {
...
+    token.nonEmpty && token.length > len && {
+      val (hmac, timestamp) = (token.dropRight(len), token.takeRight(len))
+      hmac == generateHmac(timestamp)
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/session/SessionManager.scala]: maybe we shoudl use `Utils.constantTimeEquals` here?

File: core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala [hunk 6]
```diff
@@ -14,16 +14,20 @@ trait CsrfDirectives {
     * Note that this scheme can be broken when not all subdomains are protected or not using HTTPS and secure cookies,
     * and the token is placed in the request body (not in the header).
     *
+    * The risk of the previous disclaimer is mitigated by using a concatenation of a timestamp and its HMAC hash
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala]: So it's no longer a problem when subdomains are out of the user's control? If so, would be good to just remove that comment and maybe add a link to the appropriate OWASP section?

File: core/src/main/java/com/softwaremill/session/javadsl/CsrfDirectives.scala (hunk 1 above)
Comment (willemvermeer) [reply core/src/main/java/com/softwaremill/session/javadsl/CsrfDirectives.scala]: Done.

File: core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala (hunk 2 above)
Comment (willemvermeer) [reply core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala:44]: Added a deprecated warning to both java and scala implementations.

File: core/src/main/scala/com/softwaremill/session/SessionManager.scala (hunk 3 above)
Comment (willemvermeer) [reply core/src/main/scala/com/softwaremill/session/SessionManager.scala]: Thanks for this suggestion, changed as suggested.

File: core/src/main/scala/com/softwaremill/session/SessionManager.scala (hunk 4 above)
Comment (willemvermeer) [reply core/src/main/scala/com/softwaremill/session/SessionManager.scala]: Added a specific encoder/decoder to improve this section.

File: core/src/main/scala/com/softwaremill/session/SessionManager.scala (hunk 5 above)
Comment (willemvermeer) [reply core/src/main/scala/com/softwaremill/session/SessionManager.scala]: Done, thanks for pointing me to this function.

File: core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala (hunk 6 above)
Comment (willemvermeer) [reply core/src/main/scala/com/softwaremill/session/CsrfDirectives.scala]: Added a reference to OWASP and moved my addition one paragraph up.

**Comment by adamw** [comment]
Thank you! Released in 0.6.1

**Pull request merged**